### ISSUE TITLE
Use CompletabeFuture.handleAsync instead of CompletabeFuture.handle in Future.fromCompletableFuture to use provided executor?

### ISSUE THREAD

**Issue by rlindooren** [vavr-io/vavr#2495]
Hi,

First of all: thanks for you awesome work on this library 👍🏻.

This is not a bug-report since I think what I'm trying to accomplish is an exceptional case.
But perhaps it is interesting for you guys.

Version used: `0.10.2`.

I was trying to pass logging [MDC](https://logback.qos.ch/manual/mdc.html) data to executor threads that execute the Future's computation.

In order to accomplish this I have chosen to write a so called [ContextAwareExecutorService](https://gist.github.com/rlindooren/df53abb3ea047caa7f725b4929dc3341#file-contextawareexecutorservice-java) which basically wraps a cached thread pool and synchronizes the thread bound MDC data between threads.

Then, next to VAVR's Future, I created a [ContextAwareFuture](https://gist.github.com/rlindooren/df53abb3ea047caa7f725b4929dc3341#file-contextawarefuture-java) that calls VAVR's `Future` with my own executor.

For `Try` and `Callable` this worked as expected.
It however did not work when using `Future.fromCompletableFuture(Executor executor, CompletableFuture<T> future)` 

I think what causes this, is that the provided executor is not used to execute the future computation.

The provided executor is used internally by VAVR to `run`, which asynchronously waits on the future computation to complete.
The computation however is then still executed by the default executor (e.g. `ForkJoinPool.commonPool()`).
_this is at least what I *think* is happening_

https://github.com/vavr-io/vavr/blob/a8f4592eda322dc594f4516a8a235cf874d1f458/vavr/src/main/java/io/vavr/concurrent/Future.java#L333

How I got around this:

```java
public static <T> Future<T> fromCompletableFuture(CompletableFuture<T> future) {
    // When using this method the MDC will not be passed on the executor thread
    // most likely because future.handle is used instead of future.handleAsync(...executorService...)
    //return Future.fromCompletableFuture(executorService, future);

    // Using `get` feels counter intuitive as it is blocking
    // but it is executed asynchronously on a separate executor thread
    // and this approach allows the MDC thread data to be passed on.
    return fromTry(Try.of(future::get).recoverWith(error -> Try.failure(error.getCause())));
}
```

https://gist.github.com/rlindooren/df53abb3ea047caa7f725b4929dc3341#file-contextawarefuture-java-L33

It is possible that I'm misunderstanding this. Sorry if that's the case.

Regards,

Ricardo

**Comment by danieldietrich** [comment]
Hi, thanks for your question!

In fact, the run() call below just registers a handler on the CompletableFuture's CompletionStage.

<img width="823" alt="Screenshot 2019-08-21 at 10 51 30" src="https://user-images.githubusercontent.com/743833/63417902-72696800-c402-11e9-8dd7-100c21932235.png">

Since the given Executor is stored internally in the Future instance, subsequent operations, like onSuccess and onFailure handlers are executed asynchronously on that Executor.

If you ensure that your CompletableFuture also uses your ExecutorService, all should be fine. I think it isn't worth the overhead of creating a heavy weight thread for just setting the result of the Future when the CompletableFuture finished.

**Comment by rlindooren** [comment]
Alright, thanks for your feedback.

**Comment by danieldietrich** [comment]
One more thing: `get()` generally is not a good idea. I will think about how I can improve the integration of CompletableFuture in general. Maybe we need to provide additional factory methods that reflect the behavior of the CompletionStage.handle(*) and .handleAsync(*) methods.

I will create a separate issue since we do not change the behavior of the existing method. 

Thx
